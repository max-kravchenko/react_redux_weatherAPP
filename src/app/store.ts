import { configureStore, combineReducers } from '@reduxjs/toolkit';
import weatherSliceReducer from '../features/weatherSlice';

const rootReducer = combineReducers({
  weatherSliceReducer,
});

export const store = configureStore({
  reducer: rootReducer,
});

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = typeof store; 
export type AppDispatch = AppStore['dispatch'];
