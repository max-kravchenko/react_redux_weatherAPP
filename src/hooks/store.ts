import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { AppDispatch, RootState } from '../app/store';

export const useCustomDispatch = () => useDispatch<AppDispatch>();

export const useCustomSelector: TypedUseSelectorHook<RootState> = useSelector;