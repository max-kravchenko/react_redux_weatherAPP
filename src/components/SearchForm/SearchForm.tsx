import { TextField, Button } from '@mui/material';
import React, { useState } from 'react';
import { weatherSlice } from '../../features/weatherSlice';
import { useCustomDispatch, useCustomSelector } from '../../hooks/store';
import Box from '@mui/material/Box';
import { selectCurrentWeatherData } from '../../selectors/selector';
import Alert from '@mui/material/Alert';
import { fetchCurrentWeather } from '../../thunks/fetchCurrentWeather';

interface Props {
  city: string,
  setCity: (cityName: string) => void,
}
export const SearchForm = React.memo(function SearchForm({ city, setCity } : Props) {

  const dispatch = useCustomDispatch();
  const { weather, error } = useCustomSelector(selectCurrentWeatherData);

  const [errorCity, setErrorCity] = useState<boolean>(false);
  const [emptyCityErr, setEmptyCityErr] = useState<boolean>(false);

  const handleSubmit = async () => {

    if (city.trim() === '') {
      setEmptyCityErr(true);
      setCity('');
      setTimeout(() => {
        setEmptyCityErr(false);
      }, 3000);
      return;
    }

    if (weather.find((name) => name.name.toLowerCase() === city.toLowerCase())) {
      setCity('');
      setErrorCity(true);

      setTimeout(() => {
        setErrorCity(false);
      }, 3000);
      return;
    }

    try {
      dispatch(weatherSlice.actions.setLoad(true));
      dispatch(fetchCurrentWeather(city));
      setCity('');
      setErrorCity(false);
      dispatch(weatherSlice.actions.setError(''));
      dispatch(weatherSlice.actions.setLoad(true));
    } catch (error: any) {
      dispatch(weatherSlice.actions.setError(error));
    }
    finally {
      dispatch(weatherSlice.actions.setLoad(false));
    }
  };

  const handleCityEnter = (
    event: { currentTarget: { value: string; }}
  ) => setCity(event.currentTarget.value);

  const hanleSelectionClear = () => {
    dispatch(weatherSlice.actions.clear());
  };

  return (
    <Box>
      <Box
        display="flex"
        justifyContent="center"
        alignItems="center"
        component='form'
        noValidate
        autoComplete="off"
        onSubmit={async (event: {preventDefault: () => void}
        ) => {
          event.preventDefault();
          handleSubmit();
        }}
      >
        <TextField
          inputProps={{ 'data-testid': 'content-input' }}
          type="text"
          placeholder='Choose a city'
          value={city}
          onChange={handleCityEnter} 
        />
        <Button
          onClick={handleSubmit}
          color="success"
          size='large'

        >
        Show weather
        </Button>
        <Button
          onClick={hanleSelectionClear}
          color='error'
          size='small'
        >
        Clear selections
        </Button>
      </Box>

      {errorCity && (<Alert variant="filled" severity="warning">
        This city is already on the list. No need to add it twice
      </Alert>)}

      {emptyCityErr && (<Alert variant="filled" severity="warning">
        Go on, type some letters.
      </Alert>)}

      {error && (<Alert variant="filled" severity="warning">
        Enter correct city name
      </Alert>)}
    </Box>
  );
});