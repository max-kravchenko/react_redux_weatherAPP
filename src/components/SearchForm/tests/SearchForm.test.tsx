import React from 'react';
import { SearchForm } from '../SearchForm';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../../../app/store';
import userEvent from '@testing-library/user-event';

describe('SearchForm', () => {
  const setup = () => {
    const utils = render(
      <Provider store={store}>
        <SearchForm city={'Kyiv'} setCity={function (cityName: string): string {
          return cityName;
        }} />
      </Provider>);
    const input = utils.getByTestId('content-input') as HTMLInputElement;
    return {
      input,
      ...utils,
    };
  };

  it('should return city name', () => {
    const {input} = setup();
    expect(input.value).toBe('Kyiv');
  });

  it('should render correctly', () => {
    expect(render( <Provider store={store}>
      <SearchForm city={'Kyiv'} setCity={function (cityName: string): string {
        return cityName;
      }} />
    </Provider>)).toMatchSnapshot();
  });  

  it('calls setCity callback handler', async () => {
    const setCity = jest.fn();

    render( <Provider store={store}>
      <SearchForm city={''} setCity={setCity} />
    </Provider>);

    await userEvent.type(screen.getByRole('textbox'), 'Rivne');

    expect(setCity).toHaveBeenCalledTimes(5);
  });
});



