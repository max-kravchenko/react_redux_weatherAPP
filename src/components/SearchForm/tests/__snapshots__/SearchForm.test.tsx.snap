// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-0"
      >
        <form
          autocomplete="off"
          class="MuiBox-root css-1n2mv2k"
          novalidate=""
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                data-testid="content-input"
                id=":r1:"
                placeholder="Choose a city"
                type="text"
                value="Kyiv"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSuccess MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-root MuiButton-text MuiButton-textSuccess MuiButton-sizeLarge MuiButton-textSizeLarge css-xymarw-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Show weather
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeSmall MuiButton-textSizeSmall css-6ekzlo-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Clear selections
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-0"
    >
      <form
        autocomplete="off"
        class="MuiBox-root css-1n2mv2k"
        novalidate=""
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              data-testid="content-input"
              id=":r1:"
              placeholder="Choose a city"
              type="text"
              value="Kyiv"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-ihdtdm"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSuccess MuiButton-sizeLarge MuiButton-textSizeLarge MuiButton-root MuiButton-text MuiButton-textSuccess MuiButton-sizeLarge MuiButton-textSizeLarge css-xymarw-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Show weather
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeSmall MuiButton-textSizeSmall css-6ekzlo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Clear selections
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
